src/runtime/join.ts

```diff
@@ -497,8 +497,14 @@ export class NotScan {
   propose() { return; }
   resolveProposal() { throw new Error("Resolving a not proposal"); }
 
-  accept(multiIndex: MultiIndex, prefix, solvingFor, force?) {
-    if(!force && !this.internalVars[solvingFor.id] && this.internalVars.length || !fullyResolved(this.args, prefix)) return true;
+  accept(multiIndex: MultiIndex, prefix, solvingFor, force?, prejoin?) {
+    // if we're in the prejoin phase and this not has no args, then we need
+    // to evaluate the not to see if we should run. If we didn't do this, arg-less
+    // nots won't get evaluated during Generic Join since we're never solving for a
+    // variable that this scan cares about.
+    if((!prejoin || this.args.length)
+       && (!force && !this.internalVars[solvingFor.id] && this.internalVars.length)
+       || !fullyResolved(this.args, prefix)) return true;
```

Comment:
I find this conditional to be pretty hard to parse due to its density. It might be worth commenting about the other early acceptance cases or trying to refactor it to split it up.

This is subjective so I don't know that it's blocking.